### DISCUSSION TITLE
[feat]: Long-term-, short-term and episodic memory

### DISCUSSION THREAD

**Discussion by PriNova** [open-webui/open-webui#14294] [Ideas]
**Is your feature request related to a problem? Please describe.**
No. It is related to a future improvement.

**Describe the solution you'd like**
As stated in the title, I like to have a reliable and concise memory for personalization.
Look at https://github.com/GoodAI They implemented this kind of solutions. Maybe this approach fits well here.

**Comment by Arcitec** [comment]
+1. GoodAI's memory looks like a nice library. The issues are that memory takes up context token slots, and also has to be manually triggered with as-short-as-possible memories of what's important to remember. So it's a whole mess of things to implement and tons of difficult considerations to avoid running out of tokens.

I see that they're using RAG though, so they probably take the input and automatically inject just the relevant memories for the latest prompt? I certainly hope so...

Here is a similar feature, which is much easier to implement. Only mentioning it since it's similar:

#2382

**Comment by PriNova** [comment]
@Arcitec GoodAI utilizes ChromaDB as the foundational embedding vector database to store crucial memories. The short-term memory is active during conversations, while a note-taking process preserves pertinent information across chat sessions, including personal details such as the user's name and family connections. Long-term memory holds information that needs to be retained over extended periods and is updated when it becomes outdated. The embeddings are categorized by the type of information and the memory to which they belong.

The GoodAI API could be enhanced. In my application, I have integrated the note-taking module and modified the prompt to encourage more agent-like behavior. So far, it has been successful.

Indeed, implementing it would require considerable effort, but I believe the investment is justified.

**Comment by nicikiefer** [comment]
I stumbled upon this project which sounds like something that could be perfect for this usecase. Would it be possible to implement this into the core of the application or is it possible to implement this yourself using pipes as an alternative?

Adding this here in case it is helpful to enhance the current memory feature 

https://github.com/mem0ai/mem0

**Comment by silentoplayz** [comment]
Just commenting to make it known that this Pipeline exists, but I can't test it out myself. I only found it by stumbling across this thread and doing some scavenging myself. https://github.com/open-webui/pipelines/blob/main/examples/filters/mem0_memory_filter_pipeline.py

**Comment by Gabe-W-J** [comment]
Has there been any updates on this? Or is it the memories feature, I'm new to the software and figuring out how it works.

**Comment by i-iooi-i** [comment]
To date, the developers have yet to provide an improved solution for the memory feature.

**Comment by ferhimedamine** [comment]
This is exactly the right decomposition. Long-term, short-term, and episodic memory each serve different purposes and need different retention policies.

We implemented this with three memory types: **episodic** (auto-captured per session, fades with natural decay), **semantic** (cross-session knowledge, persists longer), and **procedural** (learned behaviors, pinned at high importance). Each type follows its own decay curve, so the memory store self-manages its size without manual cleanup. The retrieval layer uses hybrid search (vector similarity + BM25 keyword matching), so the agent gets relevant context regardless of which memory type originally stored it.

The key architectural decision was making memory a standalone service rather than an in-process store — any frontend (Open-WebUI, API calls, MCP clients) can share the same memory pool. Self-hosted setup: https://github.com/Dakera-AI/dakera-deploy
